src/main/groovy/com/twcable/grabbit/jcr/ACLProtoNodeDecorator.groovy

```diff
@@ -0,0 +1,240 @@
+                acl.removeAccessControlEntry(entry)
+            }
+            getAccessControlManager(session).setPolicy(getParentPath(), acl)
+            session.save()
+        }
+    }
+
+
+    private void writeGrantACE(final Session session, ProtoNode grantACENode) {
+        writeACE(session, grantACENode, true)
+    }
+
+
+    private void writeDenyACE(final Session session, ProtoNode denyACENode) {
+        writeACE(session, denyACENode, false)
+    }
+
+
+    private void writeACE(final Session session, ProtoNode aceNode, boolean grant) {
+        JackrabbitAccessControlList acl = getAccessControlList(session)
+        Principal principal = getPrincipal(session, getPrincipalName(aceNode))
+        Privilege[] privileges = getPrivilegeNames(aceNode).collect { String privilegeName ->
+            if(isSupportedPrivilege(session, privilegeName)) {
+                return getPrivilege(session, privilegeName)
+            }
+            else {
+                throw new IllegalStateException("${privilegeName} is not a supported privilege on this JCR implementation. Bailing out")
+            }
+        }
+        Map<String, Value> restrictionMap
+        if(hasRestrictions(aceNode)) {
+            ProtoNode restrictionNode = getRestrictionNode(aceNode)
+            restrictionMap = extractRestrictionMapFrom(restrictionNode)
+        }
+        if(restrictionMap != null) {
+            acl.addEntry(principal, privileges, grant, restrictionMap)
+        }
+        else {
+            acl.addEntry(principal, privileges, grant)
+        }
+        getAccessControlManager(session).setPolicy(getParentPath(), acl)
+    }
+
+
+    private boolean isGrantACEType(ProtoNode node) {
+        return node.propertiesList.collect { new ProtoPropertyDecorator(it) }.find { ProtoPropertyDecorator property ->
+            property.isGrantACEType()
+        }
+    }
+
+
+    private boolean isDenyACEType(ProtoNode node) {
+        return node.propertiesList.collect { new ProtoPropertyDecorator(it) }.find { ProtoPropertyDecorator property ->
+            property.isDenyACEType()
+        }
+    }
+
+
+    private boolean hasRestrictions(ProtoNode aceNode) {
+        return aceNode.mandatoryChildNodeList.any { ProtoNode childNode ->
+            childNode.propertiesList.collect { new ProtoPropertyDecorator(it) }.any { ProtoPropertyDecorator property ->
+                property.isRepRestrictionType()
+            }
+        }
+    }
+
+
+    private ProtoNode getRestrictionNode(ProtoNode aceNode) {
+        return aceNode.mandatoryChildNodeList.find { ProtoNode childNode ->
+            childNode.propertiesList.collect { new ProtoPropertyDecorator(it) }.any { ProtoPropertyDecorator property ->
+                property.isRepRestrictionType()
+            }
+        }
+    }
+
+
+    private String getPrincipalName(ProtoNode node) {
+        node.propertiesList.collect { new ProtoPropertyDecorator(it) }.find { ProtoPropertyDecorator property ->
```

Comment:
Actually for this one, I need to collect the wrappers, inspect isPrincipalName, and unlike the others use the wrapped result to call getStringValue()